quixstreams/app.py

```diff
@@ -739,6 +776,17 @@ def run(
 
         :param dataframe: instance of `StreamingDataFrame`
         """
+        self._run(dataframe)
+
+    def _exception_handler(self, exc_type, exc_val, exc_tb):
+        fail = False
+
+        if exc_val is not None and exc_type is not SourceException:
```

Comment:
Could you explain this logic? Why is `SourceException` treated differently?

Reply:
`SourceException` is raised in the main thread when a source failed with an exception.

In that case it's safe to checkpoint the sdf before shutting down as source and sdf are not directly related, there is a topic in between.